### ISSUE TITLE
Cluster update to 1.21 and 1.22 made weave Pod CrashLoopBackOff

### ISSUE THREAD

**Issue by drz9** [rancher/rke#2943]
**RKE version:**
1.3.11

**Docker version: (`docker version`,`docker info` preferred)**
```
docker version
Client:
 Version:           20.10.7
 API version:       1.41
 Go version:        go1.13.8
 Git commit:        20.10.7-0ubuntu5~20.04.2
 Built:             Mon Nov  1 00:34:17 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          20.10.7
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.8
  Git commit:       20.10.7-0ubuntu5~20.04.2
  Built:            Fri Oct 22 00:45:53 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.5.9-0ubuntu1~20.04.4
  GitCommit:        
 runc:
  Version:          1.0.0~rc95-0ubuntu1~20.04.2
  GitCommit:        
 docker-init:
  Version:          0.19.0
  GitCommit:     
```

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
OS: 20.04.2 LTS
Kernel: 5.4.0-72-generic

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**
Bare-metal vCloud VM's

**cluster.yml file:**
I redacted IP addresses and names:

```
cluster_name: internal
kubernetes_version: v1.23.6-rancher1-1

authentication:
  strategy: x509
  sans: [x.x.x.x, kubernetes.foo.foo]

authorization:
  mode: rbac

nodes:
- address: 10.x.x.x
  hostname_override: master-01
  role:
  - controlplane
  - etcd
  ssh_key: 
  user: rke
- address: 10.x.x.x
  hostname_override: master-02
  role:
  - controlplane
  - etcd
  ssh_key: 
  user: rke
- address: 10.x.x.x
  hostname_override: master-03
  role:
  - controlplane
  - etcd
  ssh_key:
  user: rke
- address: 10.x.x.x
  hostname_override: worker-01
  role:
  - worker
  ssh_key:
  user: rke
- address: 10.x.x.x
  hostname_override: worker-02
  role:
  - worker
  ssh_key: 
  user: rke
- address: 10.x.x.x
  hostname_override: worker-03
  role:
  - worker
  ssh_key:
  user: rke

network:
  plugin: weave
  weave_network_provider:
    password: secret

services:
  etcd:
    backup_config:
      enabled: true
      interval_hours: 1
      retention: 168
      s3backupconfig:
        endpoint: secret
        region: us-east1
        access_key: secret
        secret_key: secret
        bucket_name: backup
        folder: internal/kubernetes

  kube-api:
    always_pull_images: true
    audit_log:
      enabled: true
    extra_args:
      external-hostname: kubernetes.foo.foo
      oidc-client-id: kubernetes
      oidc-groups-claim: groups
      oidc-groups-prefix: 'oidc:'
      oidc-issuer-url: https://keycloak-url/auth/realms/internal
      oidc-username-claim: preferred_username
      oidc-username-prefix: 'oidc:'
      service-account-api-audiences: api
      service-account-issuer: api
      service-account-signing-key-file: /etc/kubernetes/ssl/kube-service-account-token-key.pem
    secrets_encryption_config:
      enabled: true

dns:
  provider: coredns
  upstreamnameservers:
  - 10.x.x.x
  - 10.x.x.x
  - 10.x.x.x

ingress:
  provider: none

addons_include:
  - /etc/rke/addons/coredns-config.yml
  - /etc/rke/addons/kube-scheduler.yml
  - /etc/rke/addons/kube-etcd.yml
  - /etc/rke/addons/kube-controller.yml
  - /etc/rke/addons/kube-kubelet.yml
  - /etc/rke/addons/node-exporter.yml
```

**Steps to Reproduce:**
couldn't reproduce

**Results:**
Weave Pod crashing 

Hello guys,
we updated our internal cluster from 1.20 to 1.23 with RKE 1.3.11. We did this version by version so we went 1.20 -> 1.21 and 1.21 -> 1.22 and 1.22 -> 1.23.
On versions 1.21 and 1.22 we encountered issues with one weave pod CrashLooping. The log message was:
`Could not get Kubernetes version: the server has asked for the client to provide credentials`

When I checked that specific Pod I figured out, that this crashing Pod was using Serviceaccount Credentials like:
```
- mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-xxxxx
      readOnly: true
```

The other, working weave Pods were still using the ServiceAccount Token from the weave-net Serviceaccount in kube-system: `weave-net-token-xxxxx`

We then upgraded 1.21 to 1.22 and figured, that this specific Pod on the master Node was working again, but another Weave Pod on another master node started having this exact same behaviour.
We also saw that `rke-network-plugin-deploy-job` failed twice when rke up was running and succeeded at 3rd attempt. 
The log issue here was: 
`error: You must be logged in to the server (the server has asked for the client to provide credentials)`

As the cluster was still intact and working and we couldn't find any relevant issues on GitHub for Weave and RKE we proceeded to update to 1.23 and now we have all weave Pods working. However, they are now all using their own `kube-api-access-xxxxx `Secret for the Serviceaccount and not the one from weave-net so I guess there was some kind of change in RKE? Because the weave version is 2.8.1 the whole time and didn't change during update.
Did anyone else encounter this issue with weave when updating the cluster in this specific versions?

**Comment by drz9** [comment]
Update: it is still not working on 1.23 as the Pod was running but throwing lots of "unauthorized, failed to list pods" etc. in Logs, after restart of Weave Pod on this Node it is again CrashLooping.

**Comment by drz9** [comment]
When inspecting kube-apiserver logs, we encounter the following a lot:
`E0525 11:57:08.300871       1 authentication.go:63] "Unable to authenticate the request" err="[invalid bearer token, token audiences [\"api\"] is invalid for the target audiences [\"unknown\"]]"`

However, when testing the token and ca.crt to the API, like in this documentation: 
https://kubernetes.io/docs/tasks/run-application/access-api-from-pod/#without-using-a-proxy
We sometimes encounter 401 Unauthorized and sometimes this call works perfectly. We tested this from several (working) master nodes and their respective token and the behaviour was almost always the same.
Eventually, the CrashLooping Pod is coming back up but it is not really working, even though weave itself says status is fine.
When invoking `kubectl top nodes` the metrics for this node are shown as `<unknown>`
kube-apiserver is also logging:
`E0525 10:33:28.753003       1 available_controller.go:524] v1beta1.metrics.k8s.io failed with: Operation cannot be fulfilled on apiservices.apiregistration.k8s.io "v1beta1.metrics.k8s.io": the object has been modified; please apply your changes to the latest version and try again`
So possibly there is a dependency. 

I also found out that others had a problem with containerd and the CNI Version < 0.3.0. In our case, CNI Version is exactly 0.3.0 but we are using docker as container runtime. Not really sure if this is now an RKE issue or a Weave issue but I cannot really trust my cluster right now.

**Comment by drz9** [comment]
OK we figured out the problem. Our cluster.yml is using kube-apiserver extra_args:

```
 kube-api:
    extra_args:
      service-account-api-audiences: api
```

This argument must have been deprecated in recent versions and is now just: `api-audiences`

We found the hint in kube-apiserver logs at the beginning where it clearly said, this argument has been deprecated. kube-apiserver then defaults to the value in `service-account-issuer` as explained in https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/

After updating this argument and running `rke up` we just had to restart the Weave Pods one-by-one and they all came up working again.
